### PR TITLE
NASA Challenge_[@azmyin12]_[Space ROS meets Open 3D Engine: A Case Study On Setup Complexity, Features, and Challenges in Mobile Robot Simulation](closes #63)

### LINKED ISSUE #63: [NASA Space ROS Sim Summer Sprint Challenge 2024]: Space ROS meets Open 3D Engine: A Case Study On Setup Complexity, Features, and Challenges in Mobile Robot Simulation
### Feature description

This issue is related to the submission to NASA Space ROS Sim Summer Sprint Challenge 2024. 

Team lead Freelancer username: [azmyin12](https://www.freelancer.com/u/azmyin12)
Submission Title: Space ROS meets Open 3D Engine: A Case Study On Setup Complexity, Features, and Challenges in Mobile Robot Simulation

Related PR: TODO

**Notes to Judges**: Previously I opened an Issue to start my submission in space-ros/docker issue [#181](https://github.com/space-ros/docker/issues/181) and its related PR [#183](https://github.com/space-ros/docker/pull/183) but have resubmitted here as per the instructions of @mkhansenbot. 

**Deliverables**:  My submission will include

* A new docker image that integrates O3DE with Space-ROS via ROS 2 Gem. 
* A complete O3DE project called ```RobotSim``` that comes with two new new asset Gems, ```NasaCuriosityRoverGem``` and ```MarsYardGem```.
* Instructions on running the demo.
* A short case study detailing my experience with O3DE, its usefulness to NASA, Space ROS and major challenges that needs to be addressed [will be placed in the PR's comment section]

### Implementation considerations

This issue is related to PR #?? that introduces a new demo to demonstrate integration between Open 3D Engine and Space ROS via the ROS 2 Gem. It showcases a new test bed (on Earth) based on the JPL's Mars Yard.  A short report will be presented in the PR's comment section after completing all requirements for the submission.

pic_issue_open

going_over

Pros:  
- High fidelity rendering with physics accurate interactions (see pictures above). 
- Native ROS 2 support without the requirement for additional middle ware via the ROS 2 Gem.
- Python and Lua supported for complex environment event generation
- A dedicated Terrain tool.
- Feature rich UI for rapid prototyping.

Cons:

- High disk space (+40 GB) usage. 
- May be initially challenging to pickup without prior experience to game engines and lack of robotics specific tutorials.

### PR DESCRIPTION
**Pull request by Mechazo11**
Related Issue #63 

This PR adds the **o3de_mars_rover** demo that showcases a the Curiosity rover demo moving on a test bed inspired by the [NASA JPL's Mars Yard](https://www-robotics.jpl.nasa.gov/how-we-do-it/facilities/marsyard-iii/) testing ground. 

The provided Dockerfile is provided that builds the ```RobotSim``` O3DE project that integrates two new Assets (henceforth released with Apache 2.0 license) gems ```NasaCuriosityRoverGem``` and ```MarsYard``` gem. See the attached image for a preview of the system in action. 

submission_ready

## Short case study:

The main objective was to perform a case study on the setup complexity, asset generation pipeline, features and challenges of using Open 3D Engine for space robotics applications, specifically mobile manipulators like the Curiosity rover.  

Its first major advantage is the direct integration to the ROS 2 graph without requiring additional communication layers (see demo) thanks to the ROS 2 Gem created by [Robotec.ai](https://robotec.ai/).
 
Additionally, O3DE offers many beneficial features, including an intuitive UI, a rapidly navigable Entity Tree, reusable Prefabs, ray-traced rendering, a built-in Material Canvas, action scripting via Lua and Python, a dedicated Terrain generator, and a sophisticated lighting/camera system. These features may enable generation of scientifically accurate and complex test scenarios that may be very beneficial for Space ROS meta stack and NASA in general for developing its multi-million dollar robotic systems.

[**TODO** add pictures with one-two linear description of various feature after submission]

In summary, O3DE with the ROS 2 Gem, in my opinion, is a free alternative to Isaac Sim/Omniverse, offering similar capabilities and powered by NVIDIA's robust PhysX 5 physics engine.

[TODO complete case study after submission]
 
---

## Submission status
- ```build.sh```: Ready
- ```DockerFile```: Done
- ```run.sh```: Done
- ```README.md``` : Done
- ```Registration form```: Done
- ```Challenge form```: Ongoing
- ```Visual demo```: Ongoing

### CODE REVIEW

**Comment by Mechazo11** [comment]
Judges, the docker image, unfortunately, comes out to be over 56 GB. I think this may be why CI is failing. I was able to build the image with no issues when I allotted about 70 GB of space.

**Comment by Mechazo11** [comment]
@mkhansenbot , sir I apologize but I may not be able to finish in time. I am a resident of Baton Rouge, Louisiana and a hurricane will pass through our area in the next few hours. There is a high chance I will not be able to modify the docker image as may we loose power. I followed the official instruction and built the image starting from the base image. and I manually edited the demo_ws packages to make it my system compatible. I wanted to make the docker end to end hence created one large image in one go. I will try it now but I cannot guarantee it will be done before the deadline expires.

**Comment by Mechazo11** [comment]
@mkhansenbot, Sir, would the CI work if I begin my O3DE docker image from ```openrobotics/space_robot_demo``` ?

**Comment by Mechazo11** [comment]
@mkhansenbot sir, as you may know hurricane francine is passing through Louisiana and it has caused power outages where I live (Baton Rouge). I had begun the process of splitting the docker image so that the O3DE part built on top off 'openrobotics/space_robots' image. But due to the outtage I was not able to finish before the deadline. May I finish the Dockerfile as soon as I am able to?

**Comment by Mechazo11** [comment]
@mkhansenbot thank you for your kind words. We are all safe and sound here. I will proceed as instructed.
With best regards, @Mechazo11

**Comment by Mechazo11** [comment]
@mkhansenbot thank you very much for the opportunity. I will try my best to push the update by tomorrow night and get the build passing.

**Comment by Mechazo11** [comment]
@mkhansenbot as discussed I have modified the docker image so that it starts from the `openrobotics/space_robots_demo` image.  This image builds successfully in my machine as shown below

image

Please let me know if there is anything needed from my end.

**Comment by Mechazo11** [comment]
@mkhansenbot, Sir, even with splitting the docker file into multiple containers (build.sh is configured to build them in sequence) with `o3de_curiosity_docker` building on top of `openrobotics/space_robot_demo`, Github CI is still failing build with the error stating it has run out of space

image

Is it possible to allot more space to Github's CI? Or can a copy of the `openrobotics/space_robot_demo` be made available so that the build system can start from that image directly instead of building the three containers (moveit2, space_robot and o3de)?

Please advise what I should do? I apologize for this inconvenience.

**Comment by Mechazo11** [comment]
@mkhansenbot, Sir, even with splitting the o3de into its own container  Github CI is still failing build with the error stating it has run out of space. Over the weekend, I have tried multiple variations to this approach but all fails with the same error message
